### DISCUSSION TITLE
Variable length integers in serialization

### DISCUSSION THREAD

**Discussion by kddnewton** [ruby/prism#741] [Ideas]
Most of the integers that we store in serialization are related to offsets. For every integer we have, we use 4 bytes to represent the value. We don't need that much space for most cases.

One thing we could do instead is to use variable length integers. Protobuf describes it pretty well: https://protobuf.dev/programming-guides/encoding/#varints. This could potentially save on a fair amount of space because I would imagine most files would fit into 2^14, and almost all of the rest would fit into 2^21.

@enebo, @eregon thoughts? Would this kind of deserialization slow it down too much so that it wouldn't be worth it?

**Comment by enebo** [comment]
I have not read this particular guide but I have changed {start_offset, end_offset} to {start_offset, length} since all our APIs using this format and length usually is a small value. I then combined that with short for all that fit in 15 bits and then have highest value represent a marker to say it should be a full int (I have also played with 6bit byte values, short values and long but that is barely largely than just short+int).  This reduced the size of the serialized file by 50% or so.  In my case the speed got a tiny bit faster so the cost of reducing size was mildly beneficial to speed.  It was always doing things at byte boundaries which I think makes it a simple scheme.

I will read that document and see if I can grok it :)  It looks complicated but a lot of thought has been put into protobufs so I am optimistic.

We should try it and measure it but that I definitely think some scheme can be used without hurting perf.

**Reply by eregon** [reply]
Ah yes this is a good argument for length instead of end_offset. Should we change to that everywhere then (including deserialized nodes and original nodes)?

**Reply by enebo** [reply]
@eregon for JRuby all our methods seem to be designed around (index, length) so as it stands we are doing a little math.  I think for Nodes it is definitely my preference (and it is trivial to change in serialize.c).  TR is probably similar but if not then it would be a decision to make it compress a bit more or not.  As for C structs feeding serialization and compile.c?  It probably just depends on how compile.c consumes those offsets.

I should point out in my playground branch I am generating JRuby stuff into a jruby directory.  I am hoping we coalesce on important aspects of serialization (smaller/faster/less-processing) but we may not want to do everything the same.  This is especially true once we start considering AST-less building or just even which extra interfaces/subtypes we want.  We have a bigger discussion on best strategy for packaging too.

**Reply by eregon** [reply]
Changing to start, length would be fine for TruffleRuby, it's already what Truffle uses internally actually.

**Reply by eregon** [reply]
> We have a bigger discussion on best strategy for packaging too.

Any pointer to that?

**Comment by enebo** [comment]
Using single sign bits for continuation vs what I am doing seems like it would help on space.  I suppose my waste of space by using a single continuation byte is offset by not requiring the math that 128 bit varints use but simple int math is pretty cheap.  The reduction in bytes to process is at odds with the math to reassemble the value.  Tough to know without trying whether we would notice that overhead or not.

**Comment by eregon** [comment]
One note is while this can decrease the serialized size it won't change anything about deserialized node size e.g. for Java or Ruby nodes.

But still it seems highly worth it if we can make the serialized size smaller without impacting performance too negatively. I think we should also look at msgpack varint encoding, that may be simpler. The protobuf seems not so efficient because you need to read each 1 bit of every byte to know how long vs knowing it from the first byte read.

**Reply by eregon** [reply]
Msgpack varint: https://github.com/msgpack/msgpack/blob/master/spec.md#int-format-family
Probably more complicated than what we need since msgpack is schema less but we always know the schema.

The simplest I could think of is 1 byte for how long the integer is in number of following bytes, and then just the raw contents. Numbers smaller than 128 are probably not that common at least for offsets to justify a special handling without extra leading byte (0xxxxxx in binary like msgpack).

**Reply by eregon** [reply]
Something a bit better: the first 2 bits would determine the total size in bytes: 00 means 6-bit integer in a single byte, 01 means 14-bit integer in 2 bytes, 10 means 22-bit integer in 3 bytes, 11 means 30-bit integer in 4 bytes.
Or 11 could mean 32-bit integer in 5 bytes if we need more than 30-bit integers (1GB source), do we? Note that for Java it's not possible to use more than a 31-bit integer since max byte[] is signed int, i.e. 2^31. So it would make sense to limit YARP to 2^31 bytes files anyway I think.

**Reply by eregon** [reply]
For https://github.com/Shopify/yarp/discussions/622 it would actually be really useful to represent values bigger than 2^32 (notably values fitting in a 64-bit integer are fixnums on JRuby and TruffleRuby). So then the leading bytes for how many bytes after could be nice for that, and would allow up to 2^2048 which seems good enough for bignums in source code. Although we would probably still need to handle it if there is a bigger bignum in the source (> 4096 digits hexadecimal bignum). But that could be a different node maybe, or some optional string field, since anyway for numbers bigger than machine size the easiest way is likely to parse them from source anyway).

We could of course use 2 different encodings for those but maybe needless complexity?

**Reply by enebo** [reply]
@eregon I talked to Kevin about possibly just pounding out all the special syntax and having all numbers be base "something" string values.  So we would have to parse that back into a real number but it would work simply for bignum and it would not break on other different precision ruby impls (what are they)?  For numbers I would be happy if we just had BignumNode and FixnumNode where the first was just a String and the other was something we can readLong().    It is possible #622 fits into this but I don't care too much about getting those sizes down vs getting them simply.

As for this discussion I think we may just need to try a few different methods and hopefully one does well in time and space.

**Reply by enebo** [reply]
In reading msgpack link I believe we could use first byte for really small values AND mark how many bytes if you just decide to not support as many integer types.  So writing yp_buffer_append_u32 (used for locations++) could be highest 2 bits to indicate short (2 byte) and int (4 byte) but if if fits in 6 bits you just write the value and leave those 2 high bits 0.  So one readByte() follow by nothing, readShort(), or readInt().  A lot of elements will have a length <64.  You could get really wild the 6 bits of first byte could add 6 more bits to the short which would probably almost never write the 4 byte integer.

**Reply by eregon** [reply]
Right that's quite similar to https://github.com/Shopify/yarp/discussions/741#discussioncomment-5519996. I thought whether it's worth having a 3-bytes/22-bit integer and I think yes as it could be frequent for offsets and it could be efficient by readInt at offset-1 and &-mask the result (needed anyway to remove the 2 high bits). We could also use 11 to mean large integer and the number of bytes following are then stored in that first byte's low 6 bits. That seems pretty efficient space and time wise, and can represent numbers up to 64 bytes so 512 bits / 1024 hexa digits.

**Reply by eregon** [reply]
Re just string for bignum I agree. I think using the signed 64-bit integer as the limit for fixnum would be natural, after all that's the word size signed integer on most platforms today. CRuby would need to range check but that's unavoidable for CRuby anyway since it doesn't use a machine word sized integer for Fixnums.

**Reply by enebo** [reply]
Ok I think we are on same page for both things.  The int compression may or may not pan out but it sounds it will.

**Reply by eregon** [reply]
One idea for a tweak of the above, if values in 64-127 are frequent, we could have:
```
0bbbbbbb 0-127
10bbbbbb bbbbbbbb 14-bit integer
11ssssss ........ s's as 6-bit length + that numbers of bytes after
```

Also I remembered we discussed this at the beginning: https://github.com/Shopify/yarp/discussions/11#discussioncomment-3765145
That's even simpler, basically the above without the 14-bit integer case.
Although that's specifically designed to encode a length, not an integer in general (which are simply encoded as [nb of bytes + bytes in DER](https://en.wikipedia.org/wiki/ASN.1#Example_encoded_in_DER)).

**Reply by eregon** [reply]
From https://github.com/Shopify/yarp/issues/788#issuecomment-1501519023, 95.7% of files fit in `2**14`, 98.1% in `2**15`, 98.8% in `2**16`, 100% in `2**22`.

Something I didn't consider above is how to represent negative integers. But that could be implicit knowledge, i.e., we know offsets are always positive so we don't need to keep 1 bit for the sign of them, it can just be part of the field type.

**Reply by eregon** [reply]
With that distribution, the scheme from https://github.com/Shopify/yarp/discussions/741#discussioncomment-5519996 seems nice:
```
00bbbbbb 0-63
01bbbbbb bbbbbbbb 14-bit integer (95.7% of offsets in top 100 gems)
10bbbbbb bbbbbbbb bbbbbbbb 22-bit integer (100% of offsets in top 100 gems)
11ssssss ........ s's as 6-bit length + that numbers of bytes after, up to (2^6*8=) 512-bits integer
```

**Reply by enebo** [reply]
@eregon I was assuming the scope of this was for unsigned values.  Also, the other week I also did short with extension byte for full int (my scheme used whole byte as marker and it was approximately as good as byte or int but with bit marker both would be worse than this new scheme). 

I like your tweak idea but I don't think we will need 6 bits for number of bytes so perhaps third high bit could mark sign with 5 bits for length (or just that following bytes are meant to a signed value)?  Then it could cover all int values and not need to be a different field type.  Do we really need more than 64 bit integers for anything is another question but it sounds like a good layout.  The sign bit is a bit weird for small signed values since it would mean 11100001, 00000101 <- -5.  heh.  I don't know (I guess I don't like my own idea but will leave it since I took the time to write it).  It may be simpler to just have one scheme for signed and one for unsigned.

**Reply by eregon** [reply]
I was thinking to store negative numbers as 2-complement/native, so e.g. -5 would be `00111011` (leading 00 because fits in < 32) and the 3rd bit would tell if negative, for unsigned of course no need for a sign bit (different type of field). Not quite sure how to decode that efficiently though, so it may be a bad idea.
I need to play with it and see how to decode unsigned & signed efficiently from a ByteBuffer. I was thinking we can just use a mask to deal with the leading 2 bits but it may not be as simple.

**Reply by eregon** [reply]
I wrote up some ways to serialize & deserialize various varint schemes in Ruby at: https://gist.github.com/eregon/7940e226eaacdf116fca4666af9a0ae4
Maybe the first thing is whether we use native endian or big endian (and convert from/to native endian) in serialization.
Currently we use native endian for integers and I think that's nice, and also less overhead for both serialize & deserialize.

One thing is native endian doesn't really work for the 2-bit mask mentioned above, or rather it works but it's very different if we are in big endian or little endian. Little endian (most plaforms) actually needs to place these 2 bits as the least significant, so they are part of the first byte we read. Conversely big endian puts those bits at most significant.
The gist shows both, it works but it's non-trivial and also not so easy to manually decode little-endian (e.g. 258 is `00000010 01000001`, the `10` mark bits are in "the middle"). Also it's non-trivial to decode a 3-bytes integer on little endian, because it might not be OK to read one byte further (but OTOH we can always read one byte earlier).
Also handling sign in that form seems pretty tricky.

I also tried the [BER length encoding](https://en.wikipedia.org/wiki/X.690#Length_octets), and that's [very simple](https://gist.github.com/eregon/7940e226eaacdf116fca4666af9a0ae4#file-varint-rb-L103) if we limit ourselves to 1, 2, 4 or 8 bytes (and not 3, 5, 6, 7 bytes), because we don't need any shifting, we can just do one native read, and it supports negative numbers easily as well.
I think we should use that for our varints. Or at least start with that. It seems to make very little compromises on performance while still compressing pretty well.

Random fact, pack's `w` for BER-encoded integer seems pretty broken (`[2**16+2].pack('w') => "\x84\x80\x02"`).

**Reply by eregon** [reply]
I wonder how [BER length encoding](https://en.wikipedia.org/wiki/X.690#Length_octets) compares to @enebo's [design](https://github.com/Shopify/yarp/discussions/741#discussioncomment-5482693) which uses 15-bit `short` and keeps the highest bit to mean "use full 31-bit int". That implies serializing both the `short` and `int` as big endian to keep the MSB in the first byte (so extra byte swapping). That's really simple as well, although a better fit for offsets than varints in general. For values < 128, it's 1 (BER) vs 2 bytes, but for values < `2**16/15` it's 3 vs 2 bytes, and for larger values it's 5 vs 4 bytes. Maybe there are enough small lengths to make the < 128 case gain enough to compensate for the extra byte for larger values.
That scheme cannot represent an offset larger than 2**31 but there is probably no need for that  (Java arrays can't handle it) and we could limit offsets to that.

Probably we should try both and compare, but first we need to change endOffset to be length (= endOffset - startOffset) instead.
@enebo Do you have some existing change for that?

**Reply by enebo** [reply]
@eregon I did not make a clean change.  I only changed it for yp_serialize_{token, location,node} then played with trimming size (e.g. measured blob size running it through all files hit by a simple Rails app).  So nothing on the reading size was done which is going to be more than Loader.java.  I am working on some JRuby stuff this week and can squeak in some time for YARP but not a lot.  If you do this I am +1 since we both prefer length and we also know it will be smaller with any scheme we use.

**Reply by eregon** [reply]
I forgot to reply here, but https://github.com/ruby/yarp/pull/873 will use length instead of endOffset.
I'm now experimenting to see how much we can gain this way.

**Reply by eregon** [reply]
I used https://github.com/ruby/yarp/compare/main...eregon:yarp:varint to try the various varint strategies, I only did the serialization part and some I only mocked by writing the same number of bytes but not necessarily the correct value (since anyway we don't need to deserialize to check this).

Here are the results, `core` is truffleruby's core files, the 2nd run without argument is all `.rb` files of top-100-gems.
```
Initial, 32-bits for each uint32_t:
$ ruby serialized_size.rb core
source:        917377
serialized:   2248206
serialized/source: 2.45
$ ruby serialized_size.rb
source:      90191454
serialized: 150875080
serialized/source: 1.67

BER length encoding with endOffset:
$ ruby serialized_size.rb core
source:        917377
serialized:   1689720
serialized/source: 1.84
$ ruby serialized_size.rb
source:      90191454
serialized: 134330027
serialized/source: 1.49

For below, all with length instead of endOffset:

BER length encoding:
$ ruby serialized_size.rb core
source:        917377
serialized:   1315840
serialized/source: 1.43
$ ruby serialized_size.rb
source:      90191454
serialized:  98883526
serialized/source: 1.10

short or int:
$ ruby serialized_size.rb core
source:        917377
serialized:   1388546
serialized/source: 1.51
$ ruby serialized_size.rb
source:      90191454
serialized: 103607718
serialized/source: 1.15

2-bits prefix:
$ ruby serialized_size.rb core
source:        917377
serialized:   1167889
serialized/source: 1.27
$ ruby serialized_size.rb
source:      90191454
serialized:  82751892
serialized/source: 0.92

protobuf:
$ ruby serialized_size.rb core
source:        917377
serialized:   1162465
serialized/source: 1.27
$ ruby serialized_size.rb
source:      90191454
serialized:  82353925
serialized/source: 0.91
```
`BER length encoding` is very simple and a nice gain, but not as good as `2-bits prefix` and `protobuf` which are basically the same in terms of serialized size.

**Reply by enebo** [reply]
@eregon cool.  It feels like time to read the bytes (not standup anything but just decode them back to ints) could be a deciding factor.  Probably not but if it is easy to try then it would probably put a nail in the winner.  Memory is important for laziness as a second property we want.

**Reply by eregon** [reply]
For time to decode in Ruby, protobuf is faster than raw and probably faster than all others (as they involve reading multiple bytes at once which causes allocations or the need to use IO::Buffer) so that's good: https://github.com/ruby/yarp/discussions/836#discussioncomment-5960842
That's not the same for Java decoding though, I'll post it here when I get some measurements about that.

**Reply by eregon** [reply]
I created https://github.com/ruby/yarp/pull/927 and measured on Java.
Basically, it's a tiny bit slower with protobuf varints than raw 4-bytes integers, which intuitively kind of makes sense, it's more work and less predictable than just reading 4 bytes and reinterpret as an `int`, OTOH it's simply less bytes to read so that benefits performance to some degree too probably.
protobuf varints are such a huge gain in serialized size that I think this is highly worth it.

I tried my own implementation of decoding protobuf varint, and then tried https://github.com/protocolbuffers/protobuf/blob/v23.1/java/core/src/main/java/com/google/protobuf/BinaryReader.java#L1507 which is a little bit faster.
I used TruffleRuby's core files and measured the time to deserialize all of them.
```
Java Loader:
raw 4-bytes int vs protobuf mine vs protobuf upstream

jvm (C2):
jt ruby -rbenchmark -e 'sources = Dir.glob("src/main/ruby/truffleruby/**/*.rb").map { |file| source=File.read(file); [source, Truffle::Debug.yarp_serialize(source)] }; 100.times { p Benchmark.realtime { 10.times { sources.each { |source, serialized| Truffle::Debug.yarp_load(source, serialized) } } } }'
0.04162979500142683
0.04219745499904093
0.04144576299950131
0.04172010699949169
0.04208778300017002
vs
0.04317428000103973
0.04366730699985055
0.04323257000032754
0.04333246200076246
0.0438804910008912
vs
0.04351711600065755
0.04308311800014053
0.04294450599991251
0.04293689600126527
0.04287302500051737

jvm-ce (Graal CE):
jt -u jvm-ce ruby -rbenchmark -e 'sources = Dir.glob("src/main/ruby/truffleruby/**/*.rb").map { |file| source=File.read(file); [source, Truffle::Debug.yarp_serialize(source)] }; 100.times { p Benchmark.realtime { 10.times { sources.each { |source, serialized| Truffle::Debug.yarp_load(source, serialized) } } } }'
0.03699708500062115
0.0370449149995693
0.03713681699991866
0.03724886900090496
0.03763073499976599
vs
0.04011225900103454
0.04194237899901054
0.04002559800028394
0.04003527900022164
0.04006649899929471
vs
0.03817277000052854
0.03817133099983039
0.03813794999950915
0.03880394000043452
0.03809320999971533


serialize:
jt ruby -rbenchmark -e 'sources = Dir.glob("src/main/ruby/truffleruby/**/*.rb").map { |file| source=File.read(file); [source, Truffle::Debug.yarp_serialize(source)] }; 30.times { p Benchmark.realtime { 10.times { sources.each { |source, serialized| Truffle::Debug.yarp_serialize(source) } } } }'
vs
0.5042316790004406
0.5033410449996154
0.505833053999595
0.5033581759998924
0.5047975580000639
```
It's pretty close for raw vs protobuf upstream (the 1st and 3rd of the `vs`).

Also interesting is that deserializing is >10x faster than parsing + serializing (caveat: of course that includes JNI overhead and 2 copies for Java<->native), in other words, deserializing seems already pretty fast.

**Reply by enebo** [reply]
@eregon Looks great.  I am not terribly surprised on ratio.  If parsing was only for execution (and not representing the syntax also) this ratio would decrease but the amount I could not give an estimate.  Ruby parsing is a lot of work.  Fast deserialze does give more power to pre-compilation.

I am fairly close to running more real stuff on my old WIP branch.  Once I do I will see how much speeding up parsing would matter.  Setting up a Ruby runtime has a bunch of other overhead.
